File: java/src/marc/Record.java
package marc;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;

public class Record {
    private ByteBuffer data;
    private int baseAddress;
    private List<Field> fields;

    protected Record(byte[] buf) {
        this.data = ByteBuffer.wrap(buf);
        this.fields = null;
    }

    public synchronized Record process(boolean deep) throws ParseException {
        data.position(12);
        baseAddress = Util.atoin(data, 5);

        // TODO check for appropriate terminators
        ByteBuffer directory = data.position(24).slice().limit(baseAddress - 24 - 1);
        ByteBuffer fieldData = data.position(baseAddress).slice();

        this.fields = new ArrayList<Field>();
        while (directory.hasRemaining()) { // walk the directory
            int tag = Util.atoin(directory, 3);
            int fieldLength = Util.atoin(directory, 4);
            int fieldPosition = Util.atoin(directory, 5);

            ByteBuffer current = fieldData.position(fieldPosition).slice().limit(fieldLength - 1);
            Field field = Field.isControlField(tag) ? new ControlField(tag, current)
                    : new DataField(tag, current);

            if (deep)
                field = field.process();

            fields.add(field);
        }
        return this;
    }

    public synchronized void write(OutputStream out) throws IOException {
        if (fields == null) {
            out.write(data.array());
        } else {
            data.position(0);
            for (int i = 0; i < baseAddress; i++) {
                out.write(data.get());
            }
            for (Field field : fields) {
                field.write(out);
            }
            out.write(Constants.RECORD_TERMINATOR);
        }
    }
}

File: go/src/marc/record.go
package marc

import "io"

// Record basic representation of a MARC record
type Record struct {
	baseAddress int
	data        []byte
	fields      []Field
}

// Leader retrieves the leader (first 24 bytes) of a MARC record
func (r *Record) Leader() []byte {
	return r.data[0:24]
}

// Process processes and populates this Record's fields
func (r *Record) Process() *Record {
	nFields := (r.baseAddress - 24 - 1) / 12
	r.fields = make([]Field, nFields)
	for i := 0; i < nFields; i++ {
		p := 24 + i*12
		tag := atoi(r.data[p : p+3])
		len := atoi(r.data[p+3 : p+7])
		pos := r.baseAddress + atoi(r.data[p+7:p+12])

		if IsControlField(tag) {
			r.fields[i] = NewControlField(tag, r.data[pos:pos+len-1]).Process()
		} else {
			r.fields[i] = NewDataField(tag, r.data[pos:pos+len-1]).Process()
		}
	}
	return r
}

// Write writes rec out in MARC format
func (r *Record) Write(w io.Writer) {
	if r.fields == nil {
		w.Write(r.data)
	} else {
		w.Write(r.data[0:r.baseAddress])
		for _, field := range r.fields {
			field.Write(w)
		}
		w.Write([]byte{RecordTerminator})
	}
}
